PR autolabeling using Release Drafter config

timkimadobe opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
This PR adds the ability to automatically add labels to PRs as they are created and updated based on specified filepaths, PR titles, and branch names. It auto-generates a draft release changelog based on these labels.

Merging to main to enable GitHub Actions.

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?

My fork for this repo can be used for action testing: https://github.com/timkimadobe/aepsdk-edgeidentity-ios

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

timkimadobe committed:
Add workflow for labeling new PRs and config file required

addb commented on .github/label-pr-config.yml:
```diff
@@ -0,0 +1,84 @@
+    branch:
+      - '/fix\/.+/'
+    title:
+      - '/(?<![\w])((fix)|(fixes))(?![\w])/i'
```
should also look for `fixed` and `bug` words in title regex. Also if possible lets search for substring instead of entire string match. So we dont need to add different modifications of same word. for ex if we just look for substring fix it would work when we have fixed, fixes.

timkimadobe replied:
That grouping for the fix category makes sense! The regex has been updated to include those words as well. I was thinking of using exact word matches with whitespace boundary checks instead of substring to prevent accidental matching causing labels to be applied to unintended contexts.
Ex ([words containing these letters](https://wordfinder.yourdictionary.com/words-with-the-letter/) - in the contains field):  
`bug` -> `debug`, `buggy`, `debugging`
`fix` -> `prefix`, `affix`, `postfix`, `suffix`, etc.

If listing all the words is too cumbersome, what do you think about creating mutually exclusive word ending regex options (next example is not actual regex, but conceptual example)?
fix/es/ed -> matches fix, fixes, fixed

addb replied:
@timkimadobe I like the idea. Alternatively instead of substring we can pair your idea with `starts with` string. starts with fix, bug etc.

timkimadobe replied:
Ok! With the starts with strategy, were you thinking of only applying the label if the word is at the start of the title?
Ex for matching cases: 
"Fix for some feature"
"Fixes broken link issue"

Ex: for non-matching cases:
"Feature fix for ..."
"Redirect issue fixed"

addb replied:
I was saying starts with for a word. FIXed, FIXes etc.. which will handle the false positive cases you mentioned like debug, prefix etc..

timkimadobe replied:
Got it, that makes sense! Thanks for clarifying. I'll update the regex pattern to be able to handle the word suffix patterns we're looking for

addb commented on .github/workflows/pull-request-labeler.yml:
```diff
@@ -0,0 +1,19 @@
+- pull_request_target
+
+jobs:
+  create_release:
```
is this create_release a reserved keyword for release or is it something used to run the actions?

timkimadobe replied:
Sorry, missed this in the copy-paste; I believe this label is just a name for the job itself, it can be anything. What do you think of `apply_pr_labels`?

addb replied:
sounds good! nit: You can just say it `apply_labels`

timkimadobe replied:
Awesome! Updated job name to suggested text

emdobrin commented on .github/label-pr-config.yml:
```diff
@@ -0,0 +1,84 @@
+      - 'Type: enhancement'
+  - title: 'Bug Fixes'
+    labels:
+      - 'Type: bug'
```
+1 to that

timkimadobe replied:
That sounds good! Removed bug label from options

timkimadobe committed:
Update fix category to include more target words
Update Bug Fixes section to remove bug label reserved for issues labeling
Update job name

timkimadobe reviewed (commented):
Thanks for the review @addb and @emdobrin! Updated according to feedback, with one outstanding question on the strategy for PR title regex patterns

timkimadobe committed:
Update job name

timkimadobe reviewed (commented):
Updated job name, with continued discussion on regex strategy

timkimadobe committed:
Merge branch 'dev' into release-drafter-config-update2

kevinlind commented on .github/label-pr-config.yml:
```diff
@@ -0,0 +1,83 @@
+    branch:
+      - '/fix\/.+/'
+    title:
+      - '/(?<![\w])((fix)|(fixes)|(fixed)|(bug)|(bugfix)|(bug fix))(?![\w])/i'
```
nit: `bug fix` is unnecessary as you are already matching on `bug` and `fix`.

timkimadobe replied:
Updated to remove redundant term!

kevinlind commented on .github/label-pr-config.yml:
```diff
@@ -0,0 +1,83 @@
+    branch:
+      - '/buildscripts\/.+/'
+    title:
+      - '/(?<![\w])((buildscript)|(buildscripts))(?![\w])/i'
```
Would we want to add `CI` and `CI/CD` to this list, or are those too generic?

timkimadobe replied:
I like adding those terms! If it's too generic we can always remove them from the config or refine it to something more specific later

emdobrin commented on .github/label-pr-config.yml:
```diff
@@ -0,0 +1,83 @@
+    # Checks there is no word character (alphanumeric & underscore) before target words, or after to avoid accidental matches
+    title:
+      - '/(?<![\w])((doc)|(docs)|(documentation))(?![\w])/i'
+  - label: 'Type: fix'
```
should this be `Type: bugfix` so it will show under the Bug fixes title in release notes?

timkimadobe replied:
Yes, thank you! Updated to apply the correct label

emdobrin commented on .github/label-pr-config.yml:
```diff
@@ -0,0 +1,83 @@
+  - 'Type: fix'
+  - 'Type: maintenance'
+
+change-template: '$TITLE (#$NUMBER) @$AUTHOR'
+change-title-escapes: '\<*_&' # You can add # and @ to disable mentions, and add ` to disable code blocks.
+version-resolver:
+  major:
+    labels:
+      - 'major'
+  minor:
+    labels:
+      - 'minor'
+  patch:
+    labels:
+      - 'patch'
+  default: patch
+template: |
+  ## Changes
+
+  $CHANGES
+
+  **Full Changelog**: https://github.com/$OWNER/$REPOSITORY/compare/$PREVIOUS_TAG...$RESOLVED_VERSION
```
is this config required for the autolabeler or can we keep this just in the release-drafter?

timkimadobe replied:
I couldn't figure out how to prevent the labeler config from generating a changelog, so I was thinking if we copy the exact config from the actual release drafter we can get a preview of what the changelog will look like after the labels have been applied; the labeler job is hardcoded to only create draft releases so we don't have to worry about an accidental release

What do you think?

emdobrin replied:
Ok, I think we can give few runs with current setup and test if the release drafter overwrites the labeler draft, and in case multiple drafts are created we can revisit this.

timkimadobe reviewed (commented):
Thanks for the review @kevinlind and @emdobrin! Updated config based on feedback, with discussion on how to use the changelog settings

timkimadobe committed:
Update bugfix label and targeted words

timkimadobe committed:
Update release drafter config to match pr labeler

kevinlind reviewed (approved):
Looks good to me.

timkimadobe committed:
Update PR label action flow to auto-delete specially tagged draft release

timkimadobe committed:
Merge branch 'dev' into release-drafter-config-update2

emdobrin commented on .github/workflows/pull-request-labeler.yml:
```diff
@@ -0,0 +1,23 @@
+    - uses: release-drafter/release-drafter@v5
+      with:
+        name: v${{ github.event.inputs.tag }}
+        tag: pr-label-release-to-delete
```
nit: can we rename this to something self explanatory like `temp-tag-created-by-pull-request-labeler` or `pull-request-labeler-tag`

timkimadobe replied:
Updated to `temp-tag-created-by-pr-auto-labeler` for clarity!

emdobrin commented on .github/workflows/pull-request-labeler.yml:
```diff
@@ -0,0 +1,23 @@
+name: "Label PRs"
```
suggestion: can we rename this to pr-auto-labeling to be more consistent with the other check names?

timkimadobe replied:
Updated to suggested name!

timkimadobe committed:
Update pr label job name for consistency

timkimadobe committed:
Update temp tag name for clarity

emdobrin commented on .github/release-drafter.yml:
```diff
@@ -4,31 +4,32 @@ tag-template: 'v$RESOLVED_VERSION'
-  - 'feature'
-  - 'fix'
-  - 'maintenance'
+  - 'Type: bug'
+  - 'Type: bugfix'
+  - 'Type: buildscripts'
+  - 'Type: deprecated'
+  - 'Type: documentation'
+  - 'Type: enhancement'
+  - 'Type: feature'
+  - 'Type: fix'
+  - 'Type: maintenance'
```
should we remove the unused tags Type: bug, Type: fix, Type?

timkimadobe committed:
Update for word suffix style matching

timkimadobe committed:
Remove unused labels

emdobrin committed:
Merge branch 'dev' into release-drafter-config-update2

emdobrin committed:
Merge branch 'dev' into release-drafter-config-update2

timkimadobe committed:
Merge branch 'dev' into release-drafter-config-update2
